Refactorised tabularisation + Jupyter notebook w/ experiments.

mabilton opened the pull request:
<!-- Please mention an issue this pull request addresses. -->
Addresses #1308 - Speedup creation of lagged data.

### Summary

Hi there.

I've started some preliminary work on refactoring [`/darts/utils/data/tabularization.py`](https://github.com/unit8co/darts/blob/cab33a1b072ca35ae0a324bf9a1f52d29dbf5c2c/darts/utils/data/tabularization.py).

Here’s a basic outline of the changes I’ve made:
1.	I’ve split up the original `_create_lagged_data` function into two: one which creates the `X` and `y` arrays for training/validation (`create_lagged_features_and_labels`), and another that creates the `X` array for predicting  (`create_lagged_features`). The rationale behind this change is twofold:
	- In my personal opinion, clearly dividing the ‘logic’ of constructing the test/validation arrays and the logic of constructing the prediction arrays makes these functions more easily maintainable, testable, and understandable to users.
	- There’s some computational savings when constructing the prediction `X` array, since the `y` array is not constructed at all (note that in the current implementation of `_create_lagged_data`, `y` is still assembled even when `is_training=False`.
2. Instead of taking the union of all the specified series and then dropping `nan` columns, I instead first compute the intersection of all the dates found in each series and then lag the index of these common dates to construct the features/labels. The notable benefit of this approach is that it completely eliminates the `for` loops over the `lag` values. 
3. Unlike `_create_lagged_data`, I've made the functions I've implemented explicitly public, in case users wish to implement their own algorithm with relies on assembling lagged values.

### Other Information

From my very informal experiments, I’ve observed a ~10 fold speed-up on 'simple’ problems which involve a couple of lag values, and a ~40 fold speed-up on 'larger’ problems which involves many lag values (i.e. more than 10). If you’d like to run these experiments yourself, feel free to checkout the `tabularization_experiments.ipynb` notebook. For reference, these benchmarks were performed on a ~4 year old Dell XPS 15 laptop.

In saying all this, what I've done is still very much a work in progress. Notably:
1. I still need to write tests, although I do run 'correctness tests' in  `tabularization_experiments.ipynb` over a very large number of input parameter combinations (10k+ combos), so I'm pretty confident my implementation here is correct.
2. I still haven't quite got the prediction data generation working. More specifically, I'm a bit confused about the intended behaviour of `_create_lagged_data` when `is_training=False`. To see what I mean by this, please refer to the 'Understanding Behaviour of `_create_lagged_data` when `is_training=False`' section of `tabularization_experiments.ipynb`. For this example, why is `X = [15., 32., 61.]` and `Ts = [6]`? In particular, the `61` value contributed by `future_series` to `X` is only `-1` lags away from the `Ts = 6` value, not `-3` lags away? 
3. I notice that `_create_lagged_data` has an outer `for` loop over all of the specified `target_series`. What's the precise reason for this? Is there any assumption that we can make about the shapes of the timeseries its iterating over? In particular, would it be possible to first concatenate all the specified timeseries together, create the lagged variables, and then `np.split` the result at the end?
4. I believe that in special situations where `past_series`, `future_series`, and `target_series` are all sampled at the same frequency, using something like [`numpy.lib.stride_tricks.sliding_window_view`](https://numpy.org/devdocs/reference/generated/numpy.lib.stride_tricks.sliding_window_view.html) would probably be even faster than what I've written here, so that might be worth looking at. 

Any comments/feedback on what I've done here would be very welcome - thanks in advance for any help.

Cheers,
Matt.

eliane-maalouf commented:
Thank you for this submission. We will be looking at it shortly and get back to you as soon as possible.

mabilton commented:
Hey @eliane-maalouf , thanks for that : )

eliane-maalouf commented on darts/utils/data/tabularization.py:
```diff
@@ -161,3 +163,390 @@ def _create_lagged_data(
+            # Offset shared time indices by requested lags:
+            idx_to_get = time_idx.reshape(-1, 1) + all_lags[i]
+            X.append(_create_lagged_vals(all_series[i], idx_to_get))
+    X = np.concatenate(X, axis=1)
```
I think this part above to create X can benefit from the `create_lagged_features()` with an extra parameter indicating whether those features are being used for training or prediction (i.e., for the +1)

eliane-maalouf commented on darts/utils/data/tabularization.py:
```diff
@@ -161,3 +163,390 @@ def _create_lagged_data(
+        ("Must specify at least one of: " "`target_lags`, `past_lags`, `future_lags`."),
+    )
+    return None
+
```
I think we can skip checking the lags since this check is already done by the regression model, which will complain if no lags are provided, if I am not mistaken

eliane-maalouf reviewed (commented):
I think this refactoring could be a nice improvement to the tubularization function.
My main propositions/comments would be the following : 

- I would recommend to keep the calls to tabularization code under the function name `_create_lagged_data` and allow the specification of the function call as being done from `fit()` or from `predict()`. IMO this is less risky in terms of breaking code that is depending on this function (at least the existing unittest should still pass). 
- Currently, only `fit()` is using the tabularization code being refactored. `predict()` is redoing a similar process (without calling `_create_lagged_data`) and allowing to add previous predictions when the prediction horizon `n > output_chunk_length`. It might be nice to unify the features generation across the different functions. It would make the code easier to maintain indeed. As for the `is_training` argument is currently used in `shap_explainer.py` which also needs to build features for inference.

mabilton commented:
Hey @eliane-maalouf  - thanks for your insightful comments : ). Just a few comments/questions from me:
1. It seems to me that `_create_lagged_data` is being treated as a de-facto public method (i.e. other parts of the code base are treating `_create_lagged_data` as a function with a stable interface which is meant to be called outside of the file it is defined in). In which case, would it be worth it to just 'bite the bullet' and make `create_lagged_data` an explicitly public method, since:
    - This would clearly indicate to future users/contributors that `create_lagged_data` is meant to be called by other parts of the code base and, therefore, that changes made to that function may be breaking.
    - From my understanding, it's generally considered a 'code smell' if a private method/function is being explicitly tested/called outside of the file it's defined in; [here's a pretty good discussion](https://softwareengineering.stackexchange.com/questions/380287/why-is-unit-testing-private-methods-considered-as-bad-practice) about this point on StackExchange.
   - Searching over the code base, it appears that only `shap_explainer.py`, `catboost_model.py`, `gradient_boost_model.py`,  `regression_model.py`, and `test_regression_models.py` explicitly call `_create_lagged_data`, so I don't suspect it would be a huge job to change these references.
2. I didn't realise that `RegressionModel.predict` was basically 're-implementing' prediction feature generation; when I have time, I'll look over this and get back to you.
3. Fair enough point on 'unifying' training and prediction feature generation under a single function using the `is_training` argument. One option which gives us the 'best of both worlds' might be to have an additionally helper function that calls either `create_lagged_features_and_labels` or `create_lagged_features` depending on whether `is_training`.

eliane-maalouf commented:
Hello @mabilton , to follow up on this PR, 
- we propose to leave out the predict part for another PR, this way this PR can remain limited in scope for the tabularization function. We are creating a separate issue for the unification of fit() and predict() feature generation.
- Indeed only few classes are using `create_lagged_data`. So maybe the best way forward is that you supplement this PR with the changes using your functions and make sure that the current unittest pass correctly so we can validate that the proposed changes are not breaking the current state of the library. 
- concerning making `create_lagged_data` a public method, I think I will ask @hrzn for his opinion on this design choice.
 In the unittests, if I am not mistaken, only the output of `create_lagged_data` is explicitly tested against an expected output, but the inner workings of the function are not explicitly tested. In the case of the modifications you are proposing, I see that the core change is in `_find_intersecting_times` so explicitly testing the expected output of this function would still be useful IMO. WDYT?

hrzn commented:
Hi @mabilton and thanks a lot for this proposal! As far as I can tell it looks good and could be included in an upcoming release. We will need a bit more time to review thoroughly :) Few points already:

* As proposed by @eliane-maalouf, I would be in favour that you drop the old function in this PR, and replace it by yours :) This way we can truly assess what the change would look like and what the current unit tests tell us.
* I agree with you it probably makes more sense to have `create_lagged_data` public at this point. It's probably not something that we would expect users to call often, but we can expose it. You can make changes in `RegressionModels` to update the name like you propose.

Thanks again!

mabilton commented:
Hi @eliane-maalouf, @hrzn - just a quick update from me.

I've managed to implement the 'sliding window' method for equal frequency series I mentioned in my initial PR; as expected, it tends to be 2 - 3 times faster than using the 'time intersection' method when all of the specified series are of the same frequency. In implementing this, I also refactored what I previously had. Notably:
- I've split up what was `_find_time_intersection` into two new functions: `get_feature_times`, which returns the times in each series which could be used to create features, and `get_shared_times`, which just finds the temporal intersection of an arbitrary number of time indices / `TimeSeries`. Splitting up the logic of these two functions hopefully makes the code more understandable, and should make it easier to explicitly test behaviors.
- I've added a `check_inputs` argument, so that classes that have already done input checks don't need to repeat these inside of `create_lagged_data`.
- I've updated the `tabularization_experiments.ipynb` to cover more test cases. As a summary, the greatest computational gains compared to `_create_lagged_data` is for unequal frequency data when `max_samples_per_ts` is specified to a small value  (~200+ times faster). This is because, in its current form, `_create_lagged_data` computes all of the possible features and then takes the last `max_samples_per_ts` samples, whereas the newly implemented `create_lagged_data` avoids computing features it doesn't need. Also, just to be clear, I will remove `tabularization_experiments.ipynb` and `_create_lagged_data` from my PR once everything is ready - they're just there in the mean time for comparison purposes.
- Where possible, I've factored in your comments on my previous code - thanks for the help : )

In its current form, `create_lagged_data` still doesn't fully comply with what's expected by other parts of the library, so tests won't currently pass. One difficulty I'm having is that I don't quite understand how `_create_lagged_data` is meant to behave when `is_training = False`. To understand what I mean by this, it would be appreciated if you could have a quick look over the *Unexpected Behaviour of `_create_lagged_data` when `is_training=False`* section of `tabularization_experiments.ipynb`, where I go over a simple example where `_create_lagged_data` shows some (perhaps?) unexpected when `is_training = False`. Once I understand what's supposed to happen when `is_training = False`, I should be ready to alter what I have to comply with the tests.

Hopefully what I've written here makes sense - let me know if you have any questions. Once again, thanks for all your help : ).

Cheers,
Matt.

eliane-maalouf commented:
@mabilton thanks for the update, I will have a look next week.

hrzn commented:
Thanks a lot @mabilton ! This looks really promising. Please bear with us (we are slow to review atm - busy preparing v0.23). I'm hopeful we can add your refactoring and improvements into v0.24.

mabilton commented:
Hey @hrzn - no worries! All of this can definitely wait until after Christmas and the New Year; in any case, I'm currently in the process of writing up tests for what I have (and fixing bugs that I discover while doing this), so it's probably for the best that you hold off on reviewing what I have for the time being. In any case, thanks for all your hard work @hrzn and @eliane-maalouf - have a Merry Christmas and Happy New Year : ).

Cheers,
Matt.

hrzn commented:
> Hey @hrzn - no worries! All of this can definitely wait until after Christmas and the New Year; in any case, I'm currently in the process of writing up tests for what I have (and fixing bugs that I discover while doing this), so it's probably for the best that you hold off on reviewing what I have for the time being. Thanks for all your hard work @hrzn and @eliane-maalouf - have a Merry Christmas and Happy New Year : ).
> 
> Cheers,
> Matt.

Thank you @mabilton , happy new year to you too!

mabilton commented:
Hi @hrzn and @eliane-maalouf - another (hopefully final) update from me.

I've now finished writing tests; in the process of doing this, I've also made a few other changes:
1. I discovered that there was a bug in the originally implemented `_create_lagged_data`. Put simply, `_create_lagged_data` wasn't accounting for the fact that features can be created for times that extend beyond the end of a timeseries. Generally speaking, to create lagged features associated with the time `t`, we don't actually need to know the all the values of the series up to that time - we only need to know the value of the series up to time `t - min_lag`, where `min_lag` is the smallest magnitude lag value requested by the user. It's probably easier to understand this bug by checking out `tabularization_experiments.ipynb`, which I've now updated to go over a few examples related to this.
2. Speaking of `tabularization_experiments.ipynb`, I removed the 'correctness tests' where I compared the outputs of `_create_lagged_data` with my own implementation, since the original implementation was producing incorrect outputs anyways. I've kept in the benchmarks though, just to get a rough idea of the kind of performance gains achieved using the newer implementation.
3. I've included two new functions: `create_lagged_training_data` and `create_lagged_prediction_data`, which are  both simply wrapper functions around the newly implemented `create_lagged_data`. In my opinion, these functions provide more 'obvious' interfaces to the user in the sense that: 
        
   - It's more obvious which inputs are actually utilised by each function (e.g. `create_lagged_prediction_data` doesn't accept `output_chunk_length` nor `multi_models` as inputs, since these are unused when creating prediction data). 
   - Its more obvious what outputs the user should expect to receive from each function (i.e. `create_lagged_prediction_data` doesn't return a `y` at all,  instead of returning `y` as `None`).
4.  I've updated `RegressionModel` and `ShapExplainer` to use the newly implemented tabularization functions. I should note here that I moved the 'outer for loop' originally included in `_create_lagged_data` (which looped over each `TimeSeries` in a list, tabularized each one, and then concatenated them together along the timestep axis) to inside of the `RegressionModel`/`ShapExplainer` classes, since I don't quite understand why we'd want this outer loop 'by default'. I'm more than happy to move this for loop back inside `create_lagged_data` if there's a reason for it that I haven't thought of.
5. I needed to slightly modify some of the `ShapExplainer` tests; in particular, a couple of the tests were attempting to create a `RegressionModel` with a positive lag (e.g. `[2]`), which means these models would use future values to predict past values. Although this is technically possible, this seems like something we ought to prevent the user from (accidentally) doing, since these kinds of models aren't causally coherent (i.e. the future doesn't cause things in the past).
6. I moved `_add_static_covariates` into `RegressionModel`, since that was the only place it was being used. If either of you anticipate that it'll be used by other parts of the code base in the future, just let me know and I'll move it back.

Once you've both had the opportunity to look over what I've done here and approve it, I'll remove `tabularization_experiments.ipynb`, `tab_experiment_utils.py`, and `_create_lagged_data` from the PR - it'll then be ready to merge.

I appreciate that this is a pretty big PR, so I understand this will probably take a while to go over - please don't feel the need to 'rush' this through. Once again, any feedback is welcome and thanks for all your help with this PR - apologies that it's dragged on for so long. Happy New Year!

Cheers,
Matt.

mabilton commented:
Hey @hrzn, @eliane-maalouf  - just a heads up that this PR is failing at the moment due to some of the static covariates code. Just letting you know I'm working on it at the moment - it should hopefully be ready in a day or two. Thanks for your patience.

Cheers,
Matt.

hrzn commented:
Thank you for the update @mabilton. I have started to review your PR (looks really good!), and started another branch to fix the issues I was seeing with static covariates here: https://github.com/unit8co/darts/tree/mabilton-refactor-tab
I already fixed an issue with the way the function `_add_static_covariates()` is called. There's an issue remaining in some of the tests and I had the intention to also look into that, but of course if you do in the meantime it's also great.
I would propose that whoever fixes the issues first can send a status update here, if that works for you.

mabilton commented:
Hey @hrzn - sounds like a plan! I've made a bit of progress since my last comment; in particular, I've made a few notable modifications to the static covariates code:
1. The internal 'logic' of `_add_static_covariates` was coupled with the internal logic of `create_lagged_data` in the sense that the former was attempting to 'back-calculate' the number of lagged features that was created for each time series by the latter. This was the source of all the array shape errors - since the new lagged data tabularizer can produce more features per timeseries than the current implementation (see my last comment about creating features that extend beyond the end of a timeseries for more info), the number of lagged observations being calculated by `_add_static_covariates` was incorrect. Since I moved `_add_static_covariates` to inside of the `RegressionModel` class, however, this whole problem of 'back-calculating' the number of features produced for each timeseries can be entirely avoided - we just need to append on the static covariate columns to each `X` 'block' *before* concatenating these `X` blocks along the time axis (i.e. we can extract the number of features created for a particular timeseries as `X_block.shape[0]`).
2. In a similar vein, `helper_get_expected_shape_static_covs` in the test code was incorrectly calculating the expected shape; I managed to simplify this code quite a bit.
3. It appears that `_add_static_covariates` was incorrectly handling the case where timeseries contained a mixture of multi-valued static covariates (i.e. one for each component in the timeseries - this adds `n_component` additional columns to `X`) and single-value static covariates (which adds a single column to `X`). I've added two additional test cases + modified `_add_static_covariates` to handle these boundary cases.
4. With all these changes made, the only test that is currently failing is `test_static_cov_accuracy` - for whatever reason, the trained `RandomForest` perform more poorly when trained with static covariates than when not. Even more strangely, when I run the `15-static-covariates.ipynb` notebook upon which this test is based, all the results look perfectly normal. Any thoughts as to what's behind this?

Finally, before I forget, would it be wise to add/modify a test to check that the actual static covariates values being appended to `X` are correct (e.g. that they appear in the correct order)? At the moment, we're only checking that the shape of `X` is expected after appending the static covariates.

Hopefully this all makes sense - let me know if you have any questions and thanks in advanced for any help.

Cheers,
Matt.

mabilton commented:
Not sure if this helps, but this is what the predictions look like for the failing test case in `test_static_cov_accuracy`:

![test](https://user-images.githubusercontent.com/51066424/212283441-c6ee794e-5b05-42a3-954e-e12ee5270916.png)

Although the prediction with static covariates intuitively 'looks better', I guess the actual error is actually larger. Still, seems a bit strange that the model is producing a smooth prediction here when it should 'know' to produce a non-smooth one? I'll perhaps look at the feature matrices being produced a bit more closely.

mabilton commented:
Looks like I found the bug - I was incorrectly calling `_create_lagged_covariates` inside of `predict`; once I make sure all tests pass on my machine, I'll push the bug fix.

mabilton commented:
Hey @hrzn - tests are all passing now - let me know what you think about the changes : )

hrzn commented on darts/utils/data/tabularization.py:
```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
+            vals = series_i.all_values(copy=False)[
+                first_window_start_idx : first_window_end_idx + num_samples - 1, :, :
             ]
+            windows = strided_moving_window(
```
This looks really good, and I'm also thinking that at some point we could also think of exposing `stride` to users as a parameter to (indirectly) control the training set size (similar to `max_samples_per_ts`).

mabilton replied:
Great idea - I was originally intending on doing something like that, but I erred against it after seeing how big the PR was getting haha. Maybe something to add in a future PR?

hrzn replied:
Yeah, definitely for later :) I opened an issue here for us not to forget: https://github.com/unit8co/darts/issues/1487

hrzn commented on darts/utils/data/tabularization.py:
```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
+        whilst `sliding_window_view` does not.
+        2. `strided_moving_window` can only operate along a single axis, whereas `sliding_window_view` can
+        operate along multiple axes.
+    Additionally, unlike `sliding_window_view`, using `strided_moving_window` doesn't require `numpy >= 1.20.0`.
```
Looks really good, although if this there was any reason for going with `sliding_window_view` instead we could perfectly bump our Numpy requirement to `1.20.0`.
In fact, now that I see this, I think we should bump it anyway because we use `sliding_window_view` in `darts.ad` 🙀

mabilton replied:
So the key limitation of `sliding_window_view` is that it doesn't allow for stride arguments. With the being said, an easy work around would be just to call `sliding_window_view` and then extract the desired strides, but I'm not sure how much extra overhead that would add. Thoughts?

hrzn replied:
I think your current way of doing things looks like it's the best because it does allow the `stride`, which could come in handy later on if we choose to make it configurable by users. I just wrote this comment in case the Numpy version was the primary reason for your choice, to say that anyway we can (and should in fact) now require `1.20.0`. I guess the only reason for using `sliding_window_view` would be if it provides substantially better performance, otherwise `strided_moving_window()` looks best.

hrzn reviewed (commented):
Thanks and really well done @mabilton, this is really awesome work! Sorry it took this long to review it.
I have a few comments, but nothing very important (I think). One small misunderstanding is fact that `lags_past_covariates` can be 0. IMO we should not allow this as 0 is a future lag, and we are in fact preventing this in `RegressionModel`s already.

Otherwise I really like your comments (very clear & helpful), extensive tests, overall code quality, and of course the performance improvements which are made very clear from your examples in the notebook. This is super solid work. It is also a key improvement as the performance of `RegressionModel`s is an important aspect of Darts (we are also thinking of improving `historical_forecasts()` for `RegressionModel`s by leveraging the tabularization itself instead of the current for-loop over the time axis for backtesting - see https://github.com/unit8co/darts/issues/1233).

We can probably merge it pretty soon, and for sure have this part of 0.24.0. Don't forget to remove your utils file(s) and the old function from the code base.

eliane-maalouf commented on darts/models/forecasting/regression_model.py:
```diff
@@ -324,30 +324,129 @@ def _create_lagged_data(
-            past_covariates=past_covariates,
-            future_covariates=future_covariates,
-            max_samples_per_ts=max_samples_per_ts,
         )
```
wouldn't calling _add_static_covariates() better done inside the preceding for loop? in my opinion, this would help simplify the _add_static_covariates() internal logic since, if it only takes a specific target (from the input sequence) and its specific features, than one can avoid looping twice over the series inside the _add_static_covariates(), if I understood the implementation correctly. WDYT?
In the current _add_static_covariates() my assumption was that the function will receive all the features and the function should compute back everything it needs in terms of length and width of features, since I was expecting a change in the _create_lagged_data() outputs, but this might not be potentially relevant anymore with the changes you made.

eliane-maalouf replied:
It would still be necessary though to go through all the series in the input sequence once, prior, to collect the static covariates information from all of them.

mabilton replied:
Definitely - in my opinion, the static covariates would ideally be added inside of `create_lagged_data` after each 'block' has been formed, but that would probably be a bit clumsy to implement at the moment since the process of computing the static covariates requires the `n_features_in_` attribute of the `RegressionModel` object. Perhaps something to think about for a future PR?

eliane-maalouf commented on darts/models/forecasting/regression_model.py:
```diff
@@ -324,30 +324,129 @@ def _create_lagged_data(
+        elif scovs_map["covs_exist"]:
+            for i, features_i in enumerate(features):
+                vals = scovs_map["vals"][i]
+                scov_width = sum(scovs_map["sizes"].values())
```
maybe relevant in relation to my remark at line 356?

eliane-maalouf commented on darts/tests/explainability/test_shap_explainer.py:
```diff
@@ -414,7 +410,7 @@ def test_explain_with_lags_future_covariates_series_extending_into_future(self):
             lags=4,
             lags_past_covariates=[-1, -2, -3],
-            lags_future_covariates=[2],
+            lags_future_covariates=[-2],
```
I guess this test was expecting a positive lags for the future covariates (extending into the future)

eliane-maalouf commented on darts/tests/models/forecasting/test_regression_models.py:
```diff
@@ -641,100 +643,65 @@ def test_prediction_data_creation(self):
+            each timeseries (without static covariates).
+            2. The `expected_height` is the total number of rows contained across all of the
+            assembled feature matrices.
+            3. The `expected_height` is the number of columns in any one of the assembled
```
```suggestion
            3. The `expected_width` is the number of columns in any one of the assembled
```

eliane-maalouf commented on darts/tests/models/forecasting/test_regression_models.py:
```diff
@@ -745,6 +712,12 @@ def test_static_covs_addition(self):
+                "cat4": ["a"],
+            }
+        ).astype(dtype={"cat4": "category"})
 
```
thanks for adding this case as well

eliane-maalouf commented on darts/tests/models/forecasting/test_regression_models.py:
```diff
@@ -884,14 +865,44 @@ def test_static_covs_addition(self):
         target_lag = [-1]
         past_covs_lag = None
-        future_covs_lag = [-1, 3]
+        future_covs_lag = [-1, -3]
```
here it was a case to test when the output_chunk_length is smaller than the positive lag for future covariates (maybe it is not a case that happens often in practice) but in the previous implementation of _create_lagged_data it was a case that caused errors in the back computation of the length and width of the features. Would this still be relevant to test here?

eliane-maalouf commented on darts/tests/models/forecasting/test_regression_models.py:
```diff
@@ -1020,7 +1067,7 @@ def test_static_cov_accuracy(self):
             train_series_no_cov[1][:60],
         ]
-        model_no_static_cov = RandomForest(lags=period // 2, bootstrap=False)
+        model_no_static_cov = RandomForest(lags=period // 2, bootstrap=True)
```
I noticed here that bootstrap was set to True for the model_no_static_cov but was left to False for the subsequent model_static_cov, it won't be fair to compare the two. 
Furthermore, in all those tests I was only able to make RandomForest perform better, in general, when bootstrap was set to False, I would assume it is better to leave both at False. WDYT?

mabilton replied:
Good catch - I was playing around with `bootstrap` when I was debugging. It should be set back to `bootstrap=False` as it was in the original test.

eliane-maalouf reviewed (commented):
Thanks @mabilton for this work. I have mainly went through the static covariates part and added some comments.

eliane-maalouf commented:
concerning your previous question about a test to check the order of the static covariates, I think it would be a good idea to make sure that the implementation is working as expected on this regard.

mabilton commented:
Hey @hrzn, @eliane-maalouf - thank you both for all your comments. I'll try to work through them ASAP.

mabilton commented on darts/explainability/shap_explainer.py:
```diff
@@ -665,16 +666,28 @@ def _create_regression_model_shap_X(
+            past_covariates = [past_covariates] if past_covariates else None
+            future_covariates = [future_covariates] if future_covariates else None
+        X, indexes = [], []
+        for i, target in enumerate(target_series):
```
So part of the motivation behind this was so that `RegressionModel` would have access to the unconcatenated '`X` blocks' corresponding to each tabularized `TimeSeries`. By having access to the unconcatenated blocks, `_add_static_covariates` doesn't need to 'back-compute' how many lagged features were contributed to `X` by each `TimeSeries`, thereby effectively decoupling the logic of `_add_static_covariates` and `create_lagged_data`. An alternative approach to achieve this same effect would be to allow `create_lagged_data` to accept `Sequence[TimeSeries]` inputs, but have it return `Sequence[TimeSeries]` for `X` and `y`, meaning that the only repeated piece of code would be calling `np.concatenate` after calling `create_lagged_features`. What are your thoughts on this?

hrzn replied:
OK thanks for the explanation, makes sense.
One alternative, to remove the need for the caller to worry about concatenation, could be to introduce a new function `create_tabular_data(all_series: Sequence[TimeSeries]) -> np.ndarray`, which would be in charge of iterating over series and calling `create_lagged_data()` and `_add_static_covariates()` on each.
But it's not very important IMO and we can also leave it like that for the time being.

mabilton commented on darts/models/forecasting/regression_model.py:
```diff
@@ -324,19 +328,36 @@ def _create_lagged_data(
+
+        training_samples, training_labels = [], []
+        for i, target in enumerate(target_series):
+            samples_i, labels_i, _ = create_lagged_training_data(
```
Definitely - I steered clear of that for now since, admittedly, I haven't thoroughly looked through the `prediction` code and because I'm cognizant of the fact that this PR has become pretty big. With that being said, I'm definitely keen to have a go doing this if this PR is merged.

mabilton commented on darts/utils/data/tabularization.py:
```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
+    `*_series.n_components = 2`):
+        lagged_target | lagged_past_covariates | lagged_future_covariates
+    where each `lagged_*` has the following structure:
+        lag_-2_comp_1_* | lag_-2_comp_2_* | lag_-1_comp_1_* | lag_-1_comp_2_*
```
Good idea - that would definitely make the returned feature matrices a lot more interpretable to end-users.

mabilton commented on darts/utils/data/tabularization.py:
```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
+    check_inputs
+        Optionally, specifies that the `lags_*` and `series_*` inputs should be checked for validity. Caution should be
+        exercised when setting `check_inputs = False`.
+    use_moving_windows
```
`check_inputs` should be set to `False` if all the relevant input checks have been done (e.g. if the checks have been done inside `RegressionModel`). Perhaps I can just add this note to that argument?

hrzn replied:
In fact I was not thinking of `check_inputs` but rather `use_moving_windows`. Could there be cases where we want to always prevent the use of moving window even when the series have the same freq?

mabilton commented on darts/utils/data/tabularization.py:
```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
-            pad_zeros = np.zeros((1, model.n_features_in_ - features.shape[1]))
-            return np.concatenate(
-                [features, np.tile(pad_zeros, reps=(features.shape[0], 1))], axis=1
+def get_feature_times(
```
Fair point - my only (admittedly weak) reason for keeping in public is that I explicitly test/import `get_feature_times` in the testing suite, but I don't actually expect users to call it. It probably makes more sense to make it private, as you say.

hrzn replied:
Agree, I think that even if we call them from test files, so long as it's not meant to be accessed by users, it's OK (and better) to leave all these other functions in `tabularization.py` private.

mabilton commented on darts/tests/utils/tabularization/test_create_lagged_prediction_data.py:
```diff
@@ -0,0 +1,1003 @@
+        """
+        Tests that `create_lagged_prediction_data` correctly handles case when
+        `0` is included in `lags_past_covariates` and `lags_future_covariates`
+        (i.e. when we're using the values of `past_covariates` and
```
Right, that makes sense.

mabilton commented on darts/tests/explainability/test_shap_explainer.py:
```diff
@@ -377,7 +376,7 @@ def test_explain_with_lags_future_covariates_series_of_same_length_as_target(sel
         model = LightGBMModel(
             lags=4,
             lags_past_covariates=[-1, -2, -3],
-            lags_future_covariates=[2],
```
Right, that makes sense. I guess I was just coming from more of a causal inference perspective, where its generally accepted that future values can't 'cause' past values, but I guess that isn't necessarily true if one were to 'know' the future (since they'd be able to act on this information in the present). So in summary, I think you're right about this - we should allow future lags to be positive.

hrzn replied:
Exactly, for instance knowing future weather forecasts may prompt different actions and have a causal impact on targets.

mabilton commented on darts/utils/data/tabularization.py:
```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
+    of prediction from an already-trained regression model when `is_training = False`. In both cases, the time index
+    values of each generated observation is also returned.
+
+    Notes
```
Thanks haha - in my experience, there's nothing worse than trying to understand poorly documented code (especially when its code that I wrote a long time ago), so I'm trying to save my future self from as much pain as possible!

mabilton commented on darts/utils/data/tabularization.py:
```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
+    comprised only of the lagged values of `target_series`.
+
+    The shape of `X` is:
+        `X.shape = (n_observations, n_lagged_features, n_samples)`,
```
That's right - there's no fundamental reason to keep the samples axis; I just thought there would be no real harm in allowing `create_lagged_data` to be vectorised over the samples axis. 

A potentially cool application for this would be to allow users to perform Monte Carlo-style uncertainty quantification with their models. For example, although a user may start out with a single deterministic timeseries, but they can perturb this series by random noise to generate a distribution of 'pseudo-data'. They could then fit their model to each one of these perturbed series and see the resulting distribution in model parameters as a way to quantify their uncertainty in those parameters.

The results of this Monte Carlo-style analysis for 'vanilla' linear regression models wouldn't exactly be surprising (since the resulting distribution is guaranteed to be normal - see Section 13.1 of [*All of Statistics* by Larry Wasserman](https://egrcc.github.io/docs/math/all-of-statistics.pdf)), but the distribution isn't necessarily normal for regularised linear regression (I think) nor for non-linear regression models. Thoughts?

hrzn replied:
OK, nice, I like this idea, we can keep it :)
In fact, we are doing something similar (although not exactly the same thing) in `TorchForecastingModel`s, where if an input series is stochastic, we select one of its sample uniformly at random, which also allows similar Monte Carlo sampling capabilities.
Two small things maybe:
* It'd be nice to test this if possible (if you have already a test for this ignore this comment - I simply missed it :)
* We could maybe add a small note in the docstring of `RegressionModel` to describe what happens if the input series are themselves stochastic.

mabilton commented on darts/utils/data/tabularization.py:
```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
+
+    The `lags` specified for the `target_series` must all be less than or equal to -1 (i.e. one can't use the value
+    of the target series at time `t` to predict the target series at the same time `t`). Conversely, the values in
+    `lags_past_covariates` and/or `lags_future_covariates` must be less than or equal to 0 (i.e. we *are* able to
```
Right, that makes sense.

hrzn replied:
I think this part of the docstring still requires adaptation, right?

mabilton commented on darts/utils/data/tabularization.py:
```diff
@@ -165,144 +165,1181 @@ def _create_lagged_data(
+        Optionally, the lags of `past_covariates` to be used as features. Each lag value is assumed to be less than
+        or equal to 0.
+    lags_future_covariates
+        Optionally, the lags of `future_covariates` to be used as features. Each lag value is assumed to be less than
```
Not at all, haha.

mabilton commented:
Hey @hrzn, @eliane-maalouf - just letting you both know that I've read over all your very useful comments and that I'm in the process of implementing the suggested changes. In particular, I've had to make a few adjustments to the tabularization and testing code to account for the case where `lags_future_covariates` is positive. I've been a bit busy in my day-to-day life recently, so these changes may take a few more days to complete - apologies for the delay.

mabilton commented on darts/models/forecasting/regression_model.py:
```diff
@@ -324,30 +324,129 @@ def _create_lagged_data(
+
         return training_samples, training_labels
 
+    def _add_static_covariates(
```
Hmmm - definitely seeing where you're coming from. I suppose my main hesitation at the moment is that the process of computing the static covariates seems to depend a lot on the attributes of `RegressionModel`, and, from what I understand, `_add_static_covariates` is only going to be used by `RegressionModel`, which tends to give me the impression that `_add_static_covariates` is basically acting as a method for `RegressionModel` here. 

Is there somehow a way to 'decouple' computing the actual static covariates values from the `RegressionModel` object? If so, an alternative approach would be to attach the static covariates to each block inside of `create_lagged_data` as soon as each block is formed (as I noted under one of @eliane-maalouf 's suggestions).

mabilton commented:
Hey @hrzn, @eliane-maalouf . So (I think) I've managed to address pretty much all of your comments in my latest push - if I've missed something, please let me know and apologies in advance. 

There are two outstanding issues I haven't really addressed yet:
1. I don't think we've come to an agreement about where `_add_static_covariates` should be placed. In my personal opinion, it would be ideal if we could somehow refactor `_add_static_covariates`' so that it doesn't need access to the underlying `RegressionModels` object. That way, we could  think about adding static covariates to each block inside of `create_lagged_data` as soon as each block is formed. What do you guys reckon?
2. I still haven't had a chance to add an additional static covariates test, just to make sure that the actual values being appended are correct - I'll try squeeze that in before this PR is merged if I find time.

Finally, after mulling over it some more, I think that the 'moving window' method can actually be used for series of different frequencies by correctly selecting the `stride` parameter. I think that's a bit outside the scope of this PR though, but I can perhaps look to implement that alongside #1487.

Once again, any and all comments are welcome, and thanks in advance for any help - apologies for the delay in making these amendments.

Cheers,
Matt.

mabilton commented:
Looks like the `04-RNN-examples.ipynb` notebook is failing since `tensorboardX` can't be imported - any ideas what's going on here?

hrzn commented:
> Looks like the `04-RNN-examples.ipynb` notebook is failing since `tensorboardX` can't be imported - any ideas what's going on here?

Hi @mabilton and many thanks for revising your PR. Will check your updates very soon.
Regarding the issues with the CI tests, they are due to a recent release of PyTorch Lightning which was breaking a few things in our tests. I have now fixed this in `master` so the issues should be ✅ 
Thanks!

mabilton commented:
Hey @hrzn, thanks for the update + bug fix. 

I've finished amending one of the static covariates tests so that the static covariate values appended to the feature matrix are explicitly checked (as opposed to just checking that the shape of the feature matrix returned by `RegressionModel._create_lagged_data` is correct). 

To perhaps clarify what I mentioned in my previous comment about appending on the static covariate columns as soon as each feature matrix block is constructed, it might pay to have a look at `helper_get_static_covs_expected_X` in `test_regression_models.py` - this is basically a simplified implementation of the strategy I'm suggesting. Thoughts on this?

Cheers,
Matt.

hrzn reviewed (approved):
Great work @mabilton !
Only got one small comment left regarding the docstring for preventing lag=0 for past covariates.

Regarding your comment about the alternative way of adding the static covariates to the features array, as far as I can tell it makes sense 👍 I would suggest we wait for another PR and close this one first though.

Many thanks again, great stuff!

mabilton commented:
Hey @hrzn, thanks for your kind feedback. Good spot with the docstring - I've fixed that now. Thanks to both you and @eliane-maalouf for your help throughout this PR.

hrzn commented:
> Hey @hrzn, thanks for your kind feedback. Good spot with the docstring - I've fixed that now. Thanks to both you and @eliane-maalouf for your help throughout this PR.

Great thanks @mabilton :) merging now 🚀
